File: wxml_parser/Node.cpp
#include "Node.h"

namespace Web
{
    ////////////////////////////////////////////////////////////////////////
    // Node implementation
    ////////////////////////////////////////////////////////////////////////
    Node::Node() : m_children(std::vector<Node *>()),
                   m_bind_info(std::vector<std::tuple<std::string, std::string>>()) {}

    Node::~Node()
    {
        for (Node *child_node : m_children)
            delete child_node;
    }

    std::optional<std::tuple<std::string, std::string>> Node::add_child(Node *child, std::vector<std::string> *binding_events)
    {
        m_children.push_back(child);
        return std::nullopt;
    }

    size_t Node::get_num_children() const { return m_children.size(); }
    size_t Node::get_num_bind() const { return m_bind_info.size(); }
    std::string Node::get_auxiliary_data() const { return m_auxiliary_data; }
    std::string Node::get_name() const { return m_name; }

    bool Node::has_attribute(const std::vector<std::string> &attribute_names) const { return false; }
    std::optional<std::string> Node::get_attribute(const std::vector<std::string> &attribute_names) const { return std::nullopt; }

    Node *Node::get_children(long unsigned int idx) const
    {
        return (idx < m_children.size() ? m_children[idx]
                                        : nullptr);
    }

    std::tuple<std::string, std::string> Node::get_bind_info(long unsigned int idx) const
    {
        return (idx < m_bind_info.size() ? m_bind_info[idx]
                                         : std::tuple("", ""));
    }

    ////////////////////////////////////////////////////////////////////////
    // RootNode implementation
    ////////////////////////////////////////////////////////////////////////
    NodeType RootNode::type() const { return NodeType::ROOT_NODE; }
    std::string RootNode::to_string() const { return "#root {depth: 0}"; }
    int RootNode::get_depth() const { return m_depth; }

    void RootNode::add_root_child(RootNode *child)
    {
        auto return_val = Node::add_child(child);
        assert(!return_val.has_value());
        assert(child->type() == NodeType::ELEMENT_NODE);
        child->m_depth = m_depth + 1;
    }

    ////////////////////////////////////////////////////////////////////////
    // ElementWrapperNode implementation
    ////////////////////////////////////////////////////////////////////////
    ElementWrapperNode::ElementWrapperNode(const std::tuple<std::string, bool> &tag_meta_info) : RootNode()
    {
        m_name = std::get<0>(tag_meta_info);
        m_auxiliary_data = std::get<1>(tag_meta_info) ? "false" : "true";
    }

    std::optional<std::tuple<std::string, std::string>> ElementWrapperNode::add_child(Node *child, std::vector<std::string> *binding_events)
    {
        auto return_val = RootNode::add_child(child);
        assert(!return_val.has_value());
        if (child->type() == NodeType::ATTRIBUTE_NODE && binding_events)
        {
            auto it = std::find(binding_events->begin(), binding_events->end(), child->m_name);
            if (it != binding_events->end())
            {
                m_bind_info.push_back(std::tuple(child->m_name, child->m_auxiliary_data));

                return std::make_tuple(child->m_name, child->m_auxiliary_data);
            }
        }
        return std::nullopt;
    }

    NodeType ElementWrapperNode::type() const { return NodeType::ELEMENT_NODE; }

    std::string ElementWrapperNode::to_string() const
    {
        std::stringstream ss;
        ss << "#element{" << m_name << "} {depth: " << m_depth << "}";
        return ss.str();
    }

    bool ElementWrapperNode::has_end_tag() const { return m_auxiliary_data == "true"; }

    bool ElementWrapperNode::has_attribute(const std::vector<std::string> &attribute_names) const
    {
        for (Node *m_child : m_children)
        {
            if (m_child->type() == NodeType::ATTRIBUTE_NODE)
            {
                for (const std::string &attribute_name : attribute_names)
                {
                    if (m_child->m_name == attribute_name)
                        return true;
                }
            }
        }
        return false;
    }

    std::optional<std::string> ElementWrapperNode::get_attribute(const std::vector<std::string> &attribute_names) const
    {
        for (Node *m_child : m_children)
        {
            if (m_child->type() == NodeType::ATTRIBUTE_NODE)
            {
                for (const std::string &attribute_name : attribute_names)
                {
                    if (m_child->m_name == attribute_name)
                        return m_child->m_auxiliary_data;
                }
            }
        }
        return std::nullopt;
    }

    ////////////////////////////////////////////////////////////////////////
    // AttributeNode implementation
    ////////////////////////////////////////////////////////////////////////
    AttributeNode::AttributeNode(std::string attribute_name, std::string attribute_value) : Node()
    {
        m_name = attribute_name;
        m_auxiliary_data = attribute_value;
    }

    NodeType AttributeNode::type() const { return NodeType::ATTRIBUTE_NODE; }

    std::string AttributeNode::to_string() const
    {
        std::stringstream ss;
        ss << "#attribute{" << m_name << ": " << m_auxiliary_data << "}";
        return ss.str();
    }

    ////////////////////////////////////////////////////////////////////////
    // DataNode implementation
    ////////////////////////////////////////////////////////////////////////
    DataNode::DataNode(std::string data, bool is_script) : Node()
    {
        m_name = data;
        m_auxiliary_data = is_script ? "true" : "false";
    }

    NodeType DataNode::type() const { return NodeType::DATA_NODE; }

    std::string DataNode::to_string() const
    {
        std::stringstream ss;
        if (m_auxiliary_data == "true")
            ss << "script";
        ss << "data{" << m_name << "}";
        return ss.str();
    }
}

File: wxml_parser_python/Node.py
from enum import Enum
from typing import List, Tuple, Optional
from abc import ABC, abstractmethod

class NodeType(Enum):
    ROOT_NODE = 1
    ELEMENT_NODE = 2
    ATTRIBUTE_NODE = 3
    DATA_NODE = 4


class Node(ABC):
    def __init__(self):
        self.m_name: str = ""
        self.m_children: List['Node'] = []
        self.m_bind_info: List[Tuple[str, str]] = []
        self.m_auxiliary_data: str = ""
    
    def get_auxiliary_data(self) -> str:
        return self.m_auxiliary_data

    def get_name(self) -> str:
        return self.m_name
    
    def get_num_children(self) -> int:
        return len(self.m_children)

    def get_children(self, idx : int) -> Optional['Node']:
        if idx < len(self.m_children):
            return self.m_children[idx]
        return None
    
    def add_child(self, child : 'Node', binding_events: List[str] | None = None)-> Tuple[str, str] | None:
        self.m_children.append(child)
        return None
    
    def get_num_bind(self):
        return len(self.m_bind_info)
    
    def get_bind_info(self, idx : int) -> Tuple[str, str]:
        if idx < len(self.m_bind_info):
            return self.m_bind_info[idx]
        return ("", "")
    
    def has_attribute(self, attribute_names: List[str]) -> bool:
        # Dummy implementation as there's no attribute data in the base class
        return False

    def get_attribute(self, attribute_names: List[str]) -> str | None:
        # Dummy implementation as there's no attribute data in the base class
        return None

    @abstractmethod
    def type(self) -> NodeType:
        pass

    @abstractmethod
    def to_string(self) -> str:
        pass

class RootNode(Node):
    def __init__(self):
        super().__init__()
        self.m_depth = 0
    
    def type(self) -> NodeType:
        return NodeType.ROOT_NODE
    
    def to_string(self) -> str:
        return "#root {depth: 0}"

    def get_depth(self) -> int:
        return self.m_depth
    
    def add_root_child(self, child : Optional['RootNode']) -> None:
        return_val = super().add_child(child)
        assert(return_val is None)
        assert(child.type() == NodeType.ELEMENT_NODE)
        child.m_depth += 1

class ElementWrapperNode(RootNode):
    def __init__(self, tag_meta_info : Tuple[str, bool] | None):
        super().__init__()

        if tag_meta_info is not None:
            self.m_name = tag_meta_info[0]
            self.m_auxiliary_data = "False" if (tag_meta_info[1] is True) else "True"

    def type(self) -> NodeType:
        return NodeType.ELEMENT_NODE
    
    def to_string(self) -> str:
        return f"#element{{ {self.m_name} }} {{depth: {self.m_depth} }}";

    def add_child(self, child: Node, binding_events: List[str] | None = None) -> Tuple[str] | None:
        return_val = super().add_child(child, binding_events)
        assert(return_val is None)

        if (child.type() == NodeType.ATTRIBUTE_NODE and binding_events is not None):

            try:
                idx = binding_events.index(child.m_name)
                _ = binding_events[idx]  # it will be the element if found
                return (child.m_name, child.m_auxiliary_data)
            except ValueError:
                return None
        return None
    
    def has_end_tag(self) -> bool:
        return self.m_auxiliary_data == "True"
    
    def has_attribute(self, attribute_names: List[str]) -> bool:
        for child in self.m_children:
            if (child.type() == NodeType.ATTRIBUTE_NODE):
                try:
                    idx = attribute_names.index(child.m_name)
                    _ = attribute_names[idx]  # it will be the element if found
                    return True
                except ValueError:
                    continue
        return False
    
    def get_attribute(self, attribute_names: List[str]) -> str | None:
        for child in self.m_children:
            if (child.type() == NodeType.ATTRIBUTE_NODE):
                try:
                    _ = attribute_names.index(child.m_name)
                    return child.m_auxiliary_data
                except ValueError:
                    continue
        return None

    def count_num_subelements(self, element_name : str) -> int:
        count = 0
        for child in self.m_children:
            if (child.type() == NodeType.ATTRIBUTE_NODE and child.m_name == element_name):
                count += 1
        
        return count
    

class AttriuteNode(Node):

    def __init__(self, attribute_name : str, attribute_value : str):
        super().__init__()

        self.m_name = attribute_name
        self.m_auxiliary_data = attribute_value
    
    def type(self) -> NodeType:
        return NodeType.ATTRIBUTE_NODE
    
    def to_string(self) -> str:
        return f"#attribute {{ {self.m_name}: {self.m_auxiliary_data} }}"

class DataNode(Node):
        def __init__(self, data : str, is_script : bool):
            super().__init__()
            self.m_name = data
            self.m_auxiliary_data = "True" if is_script else "False"
    
        def type(self) -> NodeType:
            return NodeType.DATA_NODE
    
        def to_string(self) -> str:
            return f"#scriptdata {{ {self.m_name}}}" if self.m_auxiliary_data == "True" else f"#data {{ {self.m_name}}}"

